zebrad/tests/acceptance.rs

```diff
@@ -1273,3 +1285,75 @@ where
 
     Ok(())
 }
+
+/// Test that a failed attempt to download a genesis block will not be retried inmediatly.
+///
+/// We setup 2 nodes one connected to the other for this test.
+#[test]
+#[cfg(target_family = "unix")]
+// #[ignore]
+fn genesis_download_check() -> Result<()> {
+    // get some random ports for the 2 nodes
+    let node1_port = random_known_port();
+    let node2_port = random_known_port();
+
+    // add the seed of node2 to node1
+    let mut seeds = HashSet::new();
+    seeds.insert(format!("127.0.0.1:{}", node2_port));
+    let testdir1 = testdir()?.with_config(&mut custom_seeds_test_config(
+        seeds.clone(),
+        seeds,
+        node1_port,
+    )?)?;
+
+    // start node1
+    let mut node1 = testdir1.spawn_child(&["start"])?;
+
+    // wait to make sure node1 is ready
+    std::thread::sleep(LAUNCH_DELAY);
+
+    // start node2 with the node1 as the seed
+    let mut seeds = HashSet::new();
+    seeds.insert(format!("127.0.0.1:{}", node1_port));
+    let testdir2 = testdir()?.with_config(&mut custom_seeds_test_config(
+        seeds.clone(),
+        seeds,
+        node2_port,
+    )?)?;
+    let mut node2 = testdir2.spawn_child(&["start"])?;
+
+    // we need to wait a lot
+    std::thread::sleep(Duration::from_secs(80));
+
+    // kill both nodes
+    node1.kill()?;
+    node2.kill()?;
+
+    // we are going to testing the output in node2
+    let output2 = node2.wait_with_output()?;
+
+    // this should be the same as zebrad::components::sync::GENESIS_TIMEOUT_RETRY
+    let genesis_timeout_retry: chrono::Duration = chrono::Duration::seconds(5);
+
+    // get all lines from output2 that match "genesis". We shoud always get something as:
+    // Jun 06 10:31:51.528  INFO {zebrad="c453fbf" net="Main"}:sync: zebrad::components::sync: starting genesis block download and verify
+    // Jun 06 10:32:36.534  WARN {zebrad="c453fbf" net="Main"}:sync: zebrad::components::sync: could not download or verify genesis block, retrying e=Elapsed(())
+    // Jun 06 10:32:41.537  INFO {zebrad="c453fbf" net="Main"}:sync: zebrad::components::sync: starting genesis block download and verify
+    let lines = output2.stdout_match_line(r"genesis");
+
+    // get the time of each line
+    let mut times = Vec::new();
+    for line in lines.unwrap() {
+        let pieces: Vec<&str> = line.split("  ").flat_map(|l| l.split(' ')).collect();
+        let time = chrono::NaiveTime::parse_from_str(pieces[2], "%H:%M:%S%.3f")?;
+        times.push(time);
+    }
+
+    // make sure we always have at least 3 lines
+    assert!(times.len() >= 3);
+
+    // make sure the retry is at least GENESIS_TIMEOUT_RETRY after failure
+    assert!(times[2] >= times[1] + genesis_timeout_retry);
```

Comment:
This is a very fragile check, because it depends on the exact format of Zebra's logs. Also the test is very slow, so no-one will ever run it.

Instead, write a test in `zebrad::components::sync::tests::vectors` that passes a fake config, peers, state, and verifier to `ChainSync::new`:
https://github.com/ZcashFoundation/zebra/blob/main/zebrad/src/components/sync.rs#L203

You can use `tower::service_fn` to create fake services:
https://github.com/ZcashFoundation/zebra/pull/2254/files#diff-8853e569e32adaf7d920a17860992e7076b11d376a2844aeabe397fe28e343d8R146

Here's what you'll need to fake for each one:
- config: use the default, but pass a `Network` parameter to the test
- peers: respond to `BlocksByHash` with `Blocks(Vec::new())` (or maybe `Nil`)
- state: respond to `Depth` with `Depth(None)`
- verifier: panic - the verifier should never be called in this test

If you get any other kind of request, just panic.

@jvff can help you here, he's been doing similar things in #2254.

Reply:
If we run `0..4` requests, then they should complete within `0..15` seconds, because the first download should run immediately, and each extra one takes 5 seconds. So we should fail the test if it takes longer than 20 seconds.

This means we can run the test all the time.